File: INAPPDemo/INAPPDemo/ViewController.m
//
//  ViewController.m
//  PayByPaymentDemo
//
//  Created by lengchuanxin on 2022/8/1.
//

#import "ViewController.h"
#import "PlaceOrderViewController.h"
#import "SettingViewController.h"
#import "GPBRSA.h"
#import "GlobalDefines.h"
#import "HttpManager.h"
#import <SLDPayByPayment/SLDPayByPayment.h>
#import <Masonry/Masonry.h>
#import <MBProgressHUD/MBProgressHUD.h>

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIView *tokenBgView;
@property (weak, nonatomic) IBOutlet UITextField *tokenTextField;
@property (weak, nonatomic) IBOutlet UIView *appIdBgView;
@property (weak, nonatomic) IBOutlet UITextField *appIdTextField;
@property (weak, nonatomic) IBOutlet UIView *partnerIdBgView;
@property (weak, nonatomic) IBOutlet UITextField *partnerIdTextField;
@property (weak, nonatomic) IBOutlet UIView *deviceIdBgView;
@property (weak, nonatomic) IBOutlet UITextField *deviceIdTextField;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Setting" style:UIBarButtonItemStylePlain target:self action:@selector(settingButtonAction)];
    
    self.tokenBgView.layer.cornerRadius = 4;
    self.tokenBgView.layer.borderColor = [UIColor lightGrayColor].CGColor;
    self.tokenBgView.layer.borderWidth = 1;
    self.appIdBgView.layer.cornerRadius = 4;
    self.appIdBgView.layer.borderColor = [UIColor lightGrayColor].CGColor;
    self.appIdBgView.layer.borderWidth = 1;
    self.partnerIdBgView.layer.cornerRadius = 4;
    self.partnerIdBgView.layer.borderColor = [UIColor lightGrayColor].CGColor;
    self.partnerIdBgView.layer.borderWidth = 1;
    self.deviceIdBgView.layer.cornerRadius = 4;
    self.deviceIdBgView.layer.borderColor = [UIColor lightGrayColor].CGColor;
    self.deviceIdBgView.layer.borderWidth = 1;
    
    self.tokenTextField.text = @"02c084e52e9740e693e6a9ac9d6801f8";
    self.appIdTextField.text = [GlobalDefines appId];
    self.appIdTextField.enabled = NO;
    self.partnerIdTextField.text = [GlobalDefines partnerId];
    self.partnerIdTextField.enabled = NO;
    self.deviceIdTextField.text = @"deviceId123";
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [super touchesBegan:touches withEvent:event];
    [self.view endEditing:YES];
}

- (IBAction)paymentButtonAction:(id)sender {
    [self.view endEditing:YES];
    NSString *sign = [NSString stringWithFormat:@"iapAppId=%@&iapDeviceId=%@&iapPartnerId=%@&token=%@", self.appIdTextField.text, self.deviceIdTextField.text, self.partnerIdTextField.text, self.tokenTextField.text];
    [SLDPayByPaymentInterface paymentInViewController:self withToken:self.tokenTextField.text sign:[GPBRSA encryptString:sign privateKey:[GlobalDefines privateKey]] deviceId:self.deviceIdTextField.text resultCallback:^(NSString * _Nonnull result) {
        NSLog(@"resultCallback: %@", result);
        if ([SLDPayByPaymentConfig defaultConfig].useDefaultResultPage == NO) {
            [self showAlertWithMessage:result];
        }
        
        if ([result isEqualToString:SLDPayByPaymentResultSuccess]) {
            // Do something
        } else if ([result isEqualToString:SLDPayByPaymentResultFail]) {
            // Do something
        } else if ([result isEqualToString:SLDPayByPaymentResultPaying]) {
            // Do something
        } else if ([result isEqualToString:SLDPayByPaymentResultCancel]) {
            // Do something
        }
    }];
}

- (IBAction)placeOrderAction:(id)sender {
    PlaceOrderViewController *vc = [[PlaceOrderViewController alloc] init];
    __weak typeof(self) weakSelf = self;
    vc.confirmHandler = ^(NSString * _Nonnull amount, NSString * _Nonnull subject, NSString * _Nullable customerId) {
        NSString *merchantOrderNo = [NSString stringWithFormat:@"M%.6d%.6d", arc4random_uniform(999999), arc4random_uniform(999999)];
        NSMutableDictionary *params = [NSMutableDictionary dictionary];
        params[@"requestTime"] = @(floor([[NSDate date] timeIntervalSince1970] * 1000));
        
        NSMutableDictionary *paySceneParams = [NSMutableDictionary dictionary];
        paySceneParams[@"appId"] = [GlobalDefines appId];
        paySceneParams[@"iapDeviceId"] = weakSelf.deviceIdTextField.text;
        if (customerId.length > 0) {
            paySceneParams[@"customerId"] = customerId;
        }
        
        params[@"bizContent"] = @{
            @"merchantOrderNo": merchantOrderNo,
            @"subject": subject,
            @"totalAmount": @{
                @"currency": @"AED",
                @"amount": amount
            },
            @"paySceneCode": @"INAPP",
            @"paySceneParams": paySceneParams
        };

        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:params options:0 error:nil];
        NSString *sign = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
        
        [MBProgressHUD showHUDAddedTo:weakSelf.view animated:YES];
        [[HttpManager sharedManager] setValue:[GlobalDefines partnerId] forHTTPHeaderField:@"Partner-Id"];
        [[HttpManager sharedManager] setValue:[GPBRSA encryptString:sign privateKey:[GlobalDefines privateKey]] forHTTPHeaderField:@"sign"];
        [[HttpManager sharedManager] POST:[GlobalDefines placeOrderUrl] parameters:params success:^(id  _Nullable responseObject) {
            [MBProgressHUD hideHUDForView:weakSelf.view animated:YES];
            NSString *tokenUrl = responseObject[@"body"][@"interActionParams"][@"tokenUrl"];
            if (tokenUrl) {
                weakSelf.tokenTextField.text = tokenUrl;
                [weakSelf paymentButtonAction:nil];
            }
        } failure:^(NSError * _Nonnull error) {
            [MBProgressHUD hideHUDForView:weakSelf.view animated:YES];
            [weakSelf showAlertWithMessage:error.localizedFailureReason];
        }];
    };
    [self presentViewController:vc animated:NO completion:nil];
}

- (void)settingButtonAction {
    SettingViewController *vc = [[SettingViewController alloc] init];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)showAlertWithMessage:(NSString *)message {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Tips" message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *confirmAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        
    }];
    [alertController addAction:confirmAction];
    [self presentViewController:alertController animated:YES completion:nil];
}

@end

File: InAppDemo-Swift/InAppDemo-Swift/ViewController.swift
//
//  ViewController.swift
//  InAppDemo-Swift
//
//  Created by lengchuanxin on 2022/8/24.
//

import UIKit
import SLDPayByPayment
import MBProgressHUD

class ViewController: UIViewController {

    @IBOutlet weak var tokenBgView: UIView!
    @IBOutlet weak var tokenTextField: UITextField!
    @IBOutlet weak var appIdBgView: UIView!
    @IBOutlet weak var appIdTextField: UITextField!
    @IBOutlet weak var partnerIdBgView: UIView!
    @IBOutlet weak var partnerIdTextField: UITextField!
    @IBOutlet weak var deviceIdBgView: UIView!
    @IBOutlet weak var deviceIdTextField: UITextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Setting", style: .plain, target: self, action: #selector(settingButtonAction))
        
        tokenBgView.layer.cornerRadius = 4
        tokenBgView.layer.borderColor = UIColor.lightGray.cgColor
        tokenBgView.layer.borderWidth = 1
        appIdBgView.layer.cornerRadius = 4
        appIdBgView.layer.borderColor = UIColor.lightGray.cgColor
        appIdBgView.layer.borderWidth = 1
        partnerIdBgView.layer.cornerRadius = 4
        partnerIdBgView.layer.borderColor = UIColor.lightGray.cgColor
        partnerIdBgView.layer.borderWidth = 1
        deviceIdBgView.layer.cornerRadius = 4
        deviceIdBgView.layer.borderColor = UIColor.lightGray.cgColor
        deviceIdBgView.layer.borderWidth = 1
        
        tokenTextField.text = "02c084e52e9740e693e6a9ac9d6801f8"
        appIdTextField.text = GlobalDefines.appId()
        appIdTextField.isEnabled = false
        partnerIdTextField.text = GlobalDefines.partnerId()
        partnerIdTextField.isEnabled = false
        deviceIdTextField.text = "deviceId123"
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        view.endEditing(true)
    }

    @IBAction func paymentButtonAction(_ sender: Any) {
        view.endEditing(true)
        let token = self.tokenTextField.text ?? ""
        let deviceId = self.deviceIdTextField.text ?? ""
        let sign = "iapAppId=\(self.appIdTextField.text ?? "")&iapDeviceId=\(deviceId)&iapPartnerId=\(self.partnerIdTextField.text ?? "")&token=\(token)"
        SLDPayByPaymentInterface.payment(in: self, withToken: token, sign: GPBRSA.encryptString(sign, privateKey: GlobalDefines.privateKey()), deviceId: deviceId) { [weak self] result in
            print("resultCallback: \(result)")
            if SLDPayByPaymentConfig.default().useDefaultResultPage == false {
                self?.showAlert(with: result)
            }
            
            if result == SLDPayByPaymentResultSuccess {
                // Do something
            } else if result == SLDPayByPaymentResultFail {
                // Do something
            } else if result == SLDPayByPaymentResultPaying {
                // Do something
            } else if result == SLDPayByPaymentResultCancel {
                // Do something
            }
        }
    }
    
    @IBAction func placeOrderAction(_ sender: Any) {
        let vc = PlaceOrderViewController()
        vc.confirmHandler = { [weak self] (amount, subject, customerId) in
            guard let `self` = self else { return }
            
            let merchantOrderNo = String(format: "M%.6d%.6d", arc4random_uniform(999999), arc4random_uniform(999999))
            var params: [String: Any] = [:]
            params["requestTime"] = floor(Date().timeIntervalSince1970 * 1000)
            
            var paySceneParams: [String: Any] = [:]
            paySceneParams["appId"] = GlobalDefines.appId()
            paySceneParams["iapDeviceId"] = self.deviceIdTextField.text ?? ""
            if let customerId = customerId, customerId.count > 0 {
                paySceneParams["customerId"] = customerId
            }
            
            params["bizContent"] = [
                "merchantOrderNo": merchantOrderNo,
                "subject": subject,
                "totalAmount": [
                    "currency": "AED",
                    "amount": amount,
                ],
                "paySceneCode": "INAPP",
                "paySceneParams": paySceneParams
            ]
            
            guard let jsonData = try? JSONSerialization.data(withJSONObject: params, options: JSONSerialization.WritingOptions(rawValue: 0)) else { return }
            guard let sign = String(data: jsonData, encoding: .utf8) else { return }
            
            MBProgressHUD.showAdded(to: self.view, animated: true)
            HttpManager.shared().setValue(GlobalDefines.partnerId(), forHTTPHeaderField: "Partner-Id")
            HttpManager.shared().setValue(GPBRSA.encryptString(sign, privateKey: GlobalDefines.privateKey()), forHTTPHeaderField: "sign")
            HttpManager.shared().post(GlobalDefines.placeOrderUrl(), parameters: params) { responseObject in
                MBProgressHUD.hide(for: self.view, animated: true)
                if let dict = responseObject as? [String: Any],
                   let body = dict["body"] as? [String: Any],
                   let interActionParams = body["interActionParams"] as? [String: Any],
                   let tokenUrl = interActionParams["tokenUrl"] as? String {
                    self.tokenTextField.text = tokenUrl
                    self.paymentButtonAction(self)
                }
            } failure: { error in
                MBProgressHUD.hide(for: self.view, animated: true)
                self.showAlert(with: error.localizedDescription)
            }
        }
        present(vc, animated: false, completion: nil)
    }
    
    @objc func settingButtonAction() {
        let vc = SettingViewController()
        navigationController?.pushViewController(vc, animated: true)
    }
    
    func showAlert(with message: String) {
        let alertController = UIAlertController(title: "Tips", message: message, preferredStyle: .alert)
        let confirmAction = UIAlertAction(title: "OK", style: .default) { action in
            
        }
        alertController.addAction(confirmAction)
        present(alertController, animated: true, completion: nil)
    }
}

